### ISSUE TITLE
ERROR: Unable to write a reference to ... when building monorepo-style libraries

### ISSUE THREAD

**Issue by vbraun** [angular/angular#38876]
# 🐞 bug report

### Affected Package

@angular/compiler

### Is this a regression?

No

### Description

Monorepository-style libraries import directly from `/projects` instead of `/dist`, e.g. 
~~~
$ cat projects/lib-b/tsconfig.lib.json 
{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "paths": {
      "lib-a": [
          "projects/lib-a/src/public-api",
      ],
[...]
~~~
The advantage is that libraries do not need a separate build step. This is kinda undocumented and was initially not supported, but apparently is now:
* https://github.com/angular/angular/issues/29361
* https://github.com/angular/angular/pull/33828/files

Building an app consuming those libraries seems to work (knock on wood), but building individual libraries fails. Its not clear if building individual libraries should work in that context, but it would be nice to compile if even for type checking and testing library orders.

## 🔬 Minimal Reproduction

Minimal reproduction https://github.com/vbraun/angular-monorepo-bug with two libraries, such that lib-b imports from lib-a. Building lib-a works fine, building lib-b fails 

~~~~
$ npx ng build --project lib-b
Building Angular Package
******************************************************************************
It is not recommended to publish Ivy libraries to NPM repositories.
Read more here: https://v9.angular.io/guide/ivy#maintaining-library-compatibility
******************************************************************************

------------------------------------------------------------------------------
Building entry point 'lib-b'
------------------------------------------------------------------------------
Compiling TypeScript sources through ngc
ERROR: Unable to write a reference to SpaceComponent in /var/tmp/angular/tulip/projects/lib-a/src/lib/space.component.ts from /var/tmp/angular/tulip/projects/lib-a/src/lib/lib-a.module.ts
An unhandled exception occurred: Unable to write a reference to SpaceComponent in /var/tmp/angular/tulip/projects/lib-a/src/lib/space.component.ts from /var/tmp/angular/tulip/projects/lib-a/src/lib/lib-a.module.ts
See "/tmp/ng-TS65y3/angular-errors.log" for further details.
~~~~

## 🔥 Exception or Error

The log file says:
~~~
$ cat /tmp/ng-TS65y3/angular-errors.log
[error] Error: Unable to write a reference to SpaceComponent in /var/tmp/angular/tulip/projects/lib-a/src/lib/space.component.ts from /var/tmp/angular/tulip/projects/lib-a/src/lib/lib-a.module.ts
    at ReferenceEmitter.emit (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/imports/src/emitter.js:85:19)
    at Object.toR3Reference (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/annotations/src/util.js:249:31)
    at NgModuleDecoratorHandler._toR3Reference (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/annotations/src/ng_module.js:418:31)
    at /var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/annotations/src/ng_module.js:194:83
    at Array.map (<anonymous>)
    at NgModuleDecoratorHandler.analyze (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/annotations/src/ng_module.js:194:48)
    at TraitCompiler.analyzeTrait (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/transform/src/compilation.js:346:40)
    at analyze (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/transform/src/compilation.js:298:58)
    at _loop_1 (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/transform/src/compilation.js:320:21)
    at TraitCompiler.analyzeClass (/var/tmp/angular/tulip/node_modules/@angular/compiler-cli/src/ngtsc/transform/src/compilation.js:326:35)
~~~

## 🌍  Your Environment

**Angular Version:**
<pre><code>
$ npx ng version

     _                      _                 ____ _     ___
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | |
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | |
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 10.1.1
Node: 12.18.3
OS: linux x64

Angular: 10.1.1
... animations, cli, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.1001.1
@angular-devkit/build-angular     0.1001.1
@angular-devkit/build-optimizer   0.1001.1
@angular-devkit/build-webpack     0.1001.1
@angular-devkit/core              10.1.1
@angular-devkit/schematics        10.1.1
@ngtools/webpack                  10.1.1
@schematics/angular               10.1.1
@schematics/update                0.1001.1
ng-packagr                        10.1.0
rxjs                              6.6.3
typescript                        4.0.2
webpack                           4.44.1
</code></pre>

**Comment by JoostK** [comment]
Hi @vbraun, thanks for the excellent report and investigation into this!

The path-mapping you have references into the `.ts` file of `lib-a`, at which point the .ts files will become part of the same compilation. Because of that, any references to the components provided in `lib-a` will just be imported using a relative import. Therefore, the `generateDeepReexports` does not even have an effect, as that is only relevant when building independent targets and consuming the `.d.ts` files of upstream targets.

In the case of the CLI's library build, the library will be packaged according to APF. One of the requirements of APF is to have a single entry-point, and care is taken to prevent imports outside of the entry-point using the `rootDir` compiler option, along with `flatModuleOutFile` to instruct the Angular compiler to use a single entry-point. Having an entry-point effectively disregards the `generateDeepReexports` compiler flag:

https://github.com/angular/angular/blob/3d94919800b222e225d47dab3bbe303f1372f4ab/packages/compiler-cli/src/ngtsc/core/src/compiler.ts#L650

Because of the `rootDir` configuration, the import generation mechanism is restricted in terms of its ability to generate relative imports from `lib-b` into `lib-b`, as `lib-a` resides outside of the `rootDir`. You'd typically see TS6059 reported for this situation, and it is reported in this case but unfortunately the hard crash in the compiler masks this diagnostic message. That is something we should address.

In conclusion, it is expected—to me, at least ;)—that building your monorepo-style libraries using `ng build` is not working as you'd like, as its APF build is exactly what you don't want in a monorepo style setup.

Regarding type-checking, you can still achieve that by just running the `ngc` binary instead of `ng build`:

```bash
./node_modules/.bin/ngc -p projects/lib-b/tsconfig.lib.json --noEmit
```

this runs with the exact compiler options as configured in the tsconfig, without ng-packagr (the tool behind `ng build` for libraries) getting in the way of using a specialized tsconfig for its APF setup.

/cc @alxhub as he might have additional context to share.

**Comment by vbraun** [comment]
Thanks, calling ngc directly solves my use case!

I had seen the TS6059 before but indeed got confused by the "Unable to write a reference to" error.

**Comment by petebacondarwin** [comment]
Is it safe to close this issue as resolved then?

**Comment by jelbourn** [comment]
Closing since this seems to be resolved / working as expected.
